### PR TITLE
3775 calendar step modal

### PR DESCRIPTION
**Pull request by emitche**
Display project step modal from loan calendar

### CODE REVIEW

**Comment by emitche** [comment]
Base branch will be develop once #64 is merged.

File: app/assets/javascripts/backbone/views/project_step_view.coffee [hunk 1]
```diff
@@ -61,3 +61,4 @@ class MS.Views.ProjectStepView extends Backbone.View
     @$el.replaceWith(data)
     MS.loadingIndicator.hide()
     MS.timelineView.addBlankStep() unless @persisted || @duplicate
+    MS.CalendarView.rerenderEvents()
```
Comment (emitche) [inline app/assets/javascripts/backbone/views/project_step_view.coffee]: Are Backbone views in this project named lowerCamelCase or UpperCamelCase?

File: app/views/admin/project_steps/_calendar_step_modal_actions.slim [hunk 2]
```diff
@@ -0,0 +1,12 @@
+.actions
+  a.action.edit-step-action.btn.btn-default href="#"
+    i.fa.fa-pencil
+  - if policy(step).duplicate?
+    a.action.duplicate-step-action.btn.btn-default href="#"
```
Comment (emitche) [inline app/views/admin/project_steps/_calendar_step_modal_actions.slim]: Add additional indent?

File: app/views/admin/project_steps/_show.html.slim [hunk 3]
```diff
@@ -1,18 +1,41 @@
 .panel-body
+  - if context == 'calendar'
+    div.row
```
Comment (emitche) [inline app/views/admin/project_steps/_show.html.slim]: Additional indent?

File: app/views/admin/project_steps/_show.html.slim [hunk 4]
```diff
@@ -25,17 +48,19 @@
...
-          i.fa.fa-pencil
-        - if policy(step).duplicate?
-          a.action.duplicate-step-action href="#"
-            i.fa.fa-files-o
+      - if context == "timeline"
+        .actions
+          a.action.edit-step-action href="#"
+            i.fa.fa-pencil
+            
+          - if policy(step).duplicate? && (context == 'timeline')
```
Comment (emitche) [inline app/views/admin/project_steps/_show.html.slim]: Delete action may need to be available regardless of whether policy(step).duplicate? is true.

**Comment by emitche** [comment]
There are some outstanding bugs on this issue. I am moving to [WIP] for now.

File: app/assets/javascripts/backbone/views/calendar_step_modal_view.coffee [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+  showStepModal: (e) ->
+    calStep = e.currentTarget
+    id = @$(calStep).attr('data-step-id')
+    MS.loadingIndicator.show()
+    $.get '/admin/project_steps/' + id, context: 'calendar', (html) =>
+      @replaceContent(html)
+
+  replaceContent: (html) ->
+    @$('#calendar-step-modal').find('.modal-content').empty().append(html)
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/calendar_step_modal_view.coffee]: I think you can replace `empty().append(html)` with `html(html)`.

File: app/assets/javascripts/backbone/views/calendar_step_modal_view.coffee [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    @$('#calendar-step-modal').modal({show: true})
+    MS.loadingIndicator.hide()
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/calendar_step_modal_view.coffee]: Great job on this view! Lovely!

File: app/assets/javascripts/backbone/views/project_step_view.coffee (hunk 1 above)
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/project_step_view.coffee]: Good question. The classes are uppercase. But the timelineView above is actually an instance of a class, hence the lowercase. We should perhaps discuss this.

File: app/controllers/admin/project_steps_controller.rb [hunk 7]
```diff
@@ -16,14 +16,19 @@ def new
...
   def show
     @step = ProjectStep.find(params[:id])
     authorize @step
 
-    display_timeline(@step.project_id)
+    if request.xhr?
+      render_step_partial(:show)
+    else
+      display_timeline(@step.project_id)
+    end
```
Comment (hooverlunch) [inline app/controllers/admin/project_steps_controller.rb]: Perfect.

File: app/controllers/admin/project_steps_controller.rb [hunk 8]
```diff
@@ -48,7 +53,8 @@ def update
     # Ignore schedule shift if not successfully saved, or no subsequent steps to update.
     days_shifted = 0 unless valid && subsequent_count > 0
     render partial: "/admin/project_steps/project_step", locals: {step: @step,
-      mode: valid ? :show : :edit, days_shifted: days_shifted, subsequent_count: subsequent_count}
+      mode: valid ? :show : :edit, days_shifted: days_shifted, subsequent_count: subsequent_count,
+      context: params[:context]}
```
Comment (hooverlunch) [inline app/controllers/admin/project_steps_controller.rb]: In a case like this I think it's nicer to do:

```
render partial: "/admin/project_steps/project_step", locals: {
  step: @step,
  mode: valid ? :show : :edit, 
  days_shifted: days_shifted, 
  subsequent_count: subsequent_count,
  context: params[:context]
}
```

File: app/controllers/admin/project_steps_controller.rb [hunk 9]
```diff
@@ -164,7 +170,6 @@ def display_timeline(project_id, notice = nil)
   private
 
   def render_step_partial(mode)
-    render partial: "/admin/project_steps/project_step", locals: { step: @step, mode: mode }
+    render partial: "/admin/project_steps/project_step", locals: { step: @step, mode: mode, context: params[:context]}
```
Comment (hooverlunch) [inline app/controllers/admin/project_steps_controller.rb]: Perhaps the same here.

File: app/views/admin/calendar/index.html.slim [hunk 10]
```diff
@@ -5,10 +5,13 @@ section.calendar
...
   div#calendar.main-calendar
+  
+  = render "admin/calendar/step_modal"
 
 javascript:
   $(function(){
     new MS.Views.CalendarView({
       calEvents: #{@calEvents.to_json.html_safe}
     });
+    new MS.Views.CalendarStepModalView();
```
Comment (hooverlunch) [inline app/views/admin/calendar/index.html.slim]: You can/should put this initialization into the modal partial.

File: app/views/admin/loans/show.html.slim [hunk 11]
```diff
@@ -42,4 +32,5 @@ javascript:
       el: '.nav-tabs',
       calEvents: #{@calEvents.to_json.html_safe}
     })
+    new MS.Views.CalendarStepModalView()
```
Comment (hooverlunch) [inline app/views/admin/loans/show.html.slim]: Then you won't need to duplicate it here.

File: app/views/admin/loans/timeline/_timeline.html.slim [hunk 12]
```diff
@@ -8,7 +8,7 @@ section.timeline
       #no-steps-msg = t("project_step.no_steps")
     - else
       - @loan.project_steps.each do |step|
-        = render "admin/project_steps/project_step", step: step, mode: :show
+        = render "admin/project_steps/project_step", step: step, mode: :show, context: 'timeline'
```
Comment (hooverlunch) [inline app/views/admin/loans/timeline/_timeline.html.slim]: Always double quotes for ruby plz.

File: app/views/admin/project_steps/_project_step.html.slim [hunk 13]
```diff
@@ -1,23 +1,30 @@
-div.panel.step style="border-color: #{step.border_color}" data-step-id="#{step.id_or_temp_id}" class=step.completed_or_not
+div.panel.step [
+  style="border-color: #{step.border_color}"
+  data-step-id="#{step.id_or_temp_id}"
+  data-color="#{step.border_color}"
+  class="#{step.completed_or_not}"
+  data-context="#{context}"
```
Comment (hooverlunch) [inline app/views/admin/project_steps/_project_step.html.slim]: Nice.

File: app/views/admin/project_steps/_show.html.slim (hunk 3 above)
Comment (hooverlunch) [inline app/views/admin/project_steps/_show.html.slim]: Don't think you need it. See here when in doubt: http://slim-lang.com/

File: app/views/admin/project_steps/_show.html.slim [hunk 14]
```diff
@@ -1,73 +1,83 @@
...
-        h5.title
-          - if step.step_type_value == 'checkin'
-            .icon
-              i.fa.fa-calendar-check-o
-          - elsif step.step_type_value == 'milestone'
-            .icon
-              i.fa.fa-flag
-          .title-text = step.try(:summary)
+        - if context == 'timeline'
+            h5.title
```
Comment (hooverlunch) [inline app/views/admin/project_steps/_show.html.slim]: So this is indented one to far...

File: app/views/admin/project_steps/_show.html.slim [hunk 15]
```diff
@@ -1,73 +1,83 @@
...
+            th = t("project_step.labels.status")
+            td = step.admin_status ? t("project_step.status.#{step.admin_status}"): t("none")
+              
           tr
-            th
-              = t("project_step.labels.scheduled")
-            td style="background: #{step.scheduled_bg}"
-              = step.scheduled_date
+            th = t("project_step.labels.scheduled")
+            td style="background: #{step.scheduled_bg}" = step.scheduled_date
```
Comment (hooverlunch) [inline app/views/admin/project_steps/_show.html.slim]: Nice cleanup.

File: app/views/admin/project_steps/duplicate.html.slim [hunk 16]
```diff
@@ -1,2 +1,2 @@
 - @steps.each do |step|
-  = render "admin/project_steps/project_step", step: step, mode: step.new_record? ? :form : :show
+  = render "admin/project_steps/project_step", step: step, mode: step.new_record? ? :form : :show, context: 'timeline'
```
Comment (hooverlunch) [inline app/views/admin/project_steps/duplicate.html.slim]: Could maybe break this line using \ at end of line.

File: app/assets/stylesheets/admin/calendar/_calendar.scss [hunk 17]
```diff
@@ -171,4 +180,35 @@
       display: block;
     }
   }
+
+  #calendar-step-modal {
+    .modal-content {
+      background: none;
+      border: none;
+      box-shadow: none;
+    }
```
Comment (hooverlunch) [inline app/assets/stylesheets/admin/calendar/_calendar.scss]: A comment on this block would be appropriate. I didn't get initially why you were overriding these things.

**Comment by hooverlunch** [comment]
Remind me why the background is gray?

image

In the timeline, it is white, so I think it should be white here also...

**Comment by hooverlunch** [comment]
image

I would say this should just say 'Delete Step?'

**Comment by hooverlunch** [comment]
Overall this looks great. Excellent work. 

Note that I changed things a bit so that delete submits via ajax. The way it was, it was jumping out of calendar mode if you deleted a step.

Just a few small changes above and you can merge. If you can prioritize so that Joseph can rebase agaainst this asap that would be great.

**Pull request merged**